### PR TITLE
Update to match the 1.0 plugin api

### PR DESCRIPTION
**Pull request by Bangertm**
Here is a change to keep this working with APISpec 1.0.  The plugin api has changed to be class based.  Also based on changes in APISpec it is now more convenient for this plugin to automatically add type schemas into the spec.  So I added that too.  

**Summary of changes:** 
Moved class based plugin api
Added automatic addition of type schemas to the spec
Moved plugin code to a package
Added tests

@lafrech thoughts?

### CODE REVIEW

**Comment by timakro** [comment]
Seems like you put a lot of work into this, thanks! I will look into this tomorrow.

**Comment by lafrech** [comment]
Sorry guys, didn't get the time to look at this.

**Comment by timakro** [comment]
@Bangertm I just noticed you inherited from the main `MarshmallowPlugin`. Will this become a problem if other plugins do the same? Or if you load the original `MarshmallowPlugin` as well?

File: tests/test_apispec_oneofschema.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    assert discriminator['mapping']['flower'] ==  '#/components/schemas/Flower'
+    assert discriminator['propertyName'] == 'type'
+    assert {'$ref': '#/components/schemas/Flower'} in oneof
+    assert {'$ref': '#/components/schemas/Tree'} in oneof
+
+    assert 'Tree' in schemas
+    assert 'Flower' in schemas
+
+
+def test_rsolver_returns_none():
```
Comment (lafrech) [inline tests/test_apispec_oneofschema.py:71]: s/test_rsolver_returns_none/test_resolver_returns_none

File: apispec_oneofschema/plugin.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+def is_oneof(schema):
+    return (
+        (isinstance(schema, type) and issubclass(schema, OneOfSchema))
+        or
+        isinstance(schema, OneOfSchema)
+    )
+
+
+class MarshmallowPlugin(marshmallow.MarshmallowPlugin):
+    def init_spec(self, spec):
```
Comment (lafrech) [inline apispec_oneofschema/plugin.py:52]: The call to super will instantiate a useless instance of OpenAPIConverter in the parent class.

Not really a blocker. Just saying.

File: tests/test_apispec_oneofschema.py (hunk 1 above)
Comment (timakro) [reply tests/test_apispec_oneofschema.py:71]: Fixed in 9d9ab7036514a0e8f8fce15fa0f0cd938560dc52.

File: apispec_oneofschema/plugin.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from apispec.ext import marshmallow
+from marshmallow_oneofschema import OneOfSchema
+
+class OneofOpenAPIConverter(marshmallow.OpenAPIConverter):
+    def schema2jsonschema(self, schema):
+        if self.openapi_version.major < 3 or not is_oneof(schema):
```
Comment (lafrech) [inline apispec_oneofschema/plugin.py:23]: Doesn't OAS2 allow this kind of inheritance?

From what I remember, only `discriminator` is new to OAS3, but inheritance is possible, although loosely defined.

Should the OAS version condition only apply to the line below that adds `discriminator`?

**Comment by lafrech** [comment]
Looks like the apispec plugin API is a bit too limited for what we're trying to achieve.

We could propose here a `OneOfOpenAPIConverterMixin` with the `schema2jsonschema` override (and `is_oneof` as a static method). (I've got a thing with mixins, these days, I use mixins everywhere...)

And in apispec, we could add a class attribute (or optional init parameter ?) to pass the `OpenAPIConverter` class. This way, the user could craft a custom `MarshmallowPlugin` combining several lib mixins in an `OpenAPIConverter`.

And we could propose here a `OneOfMarshmallowPlugin` with a `OneOfOpenAPIConverter` for convenience, for the majority of users who are not in the "multiple libs customizing marshmallow plugin" case.

Thinking again, if we pass the `OpenAPIConverter` class as class attribute to the `MarshmallowPlugin`, we could need to also pass its init parameters. Maybe we're better off only accepting an `OpenAPIConverter` instance as init parameter to `MarshmallowPlugin`.

I can't look into this much further right now. I hope this helps.

One last thing. Sorry @timakro for bringing this up again, but if this feature ever makes it into apispec core, it will need to be written from scratch due to the license choice (LGPL). It's never too late for you to relicense, but if you accept external contribution, I'm afraid it will need the approval of every single contributor.

**Comment by Bangertm** [comment]
> @Bangertm I just noticed you inherited from the main `MarshmallowPlugin`. Will this become a problem if other plugins do the same? Or if you load the original `MarshmallowPlugin` as well?

My thought was this `MarshmallowPlugin` does everything that the main one does with the addition of handling `OneOfSchema` so they should just use this one instead of the main one. If both pulgins are added to spec added will overwrite the previous one and they will probably get errors about adding schemas to the spec twice.

@lafrech I think this non-ideal but more or less fine. I was thinking that it might be convenient to not have to subclass `MarshmallowPlugin` and just provide the OpenAPIConverter class desired as an initialization parameter.  `OpenAPIConverter` needs an instance of `APISpec` or else it can't add nested schemas to the spec. 

Although actually this situation is probably easier for users because all they need to change is where `MarshmallowPlugin` is imported from...

**Comment by lafrech** [comment]
Oh, right. OpenAPIConverter needs the spec object at init time so what I wrote above is wrong. And the schema name resolver. With current apispec API, it should definitely be instantiated from MarshmallowPlugin at app init time.

**Comment by timakro** [comment]
@lafrech I like to keep my projects licensed with (L)GPL because I believe in the idea that changes should be shared and others should be free to see the source and benefit from them. I would like to see this in apispec core too and I don't want to prevent the usage of this code - merging code bases is always better than forking.

If someone is implementing it in apispec core I hereby grant them the rights to use the code for [apispec-oneofschema](https://github.com/timakro/apispec-oneofschema) as well as [apispec-swaggerinherit](https://github.com/timakro/apispec-swaggerinherit) under MIT in the apispec project.

You could also contact me and I will relicense to MIT, call the project deprecated and link to the apispec project.

**Comment by lafrech** [comment]
Thanks. I knew you would. I just thought we could be kinda stuck if you merged an even minor contribution from a user that you couldn't reach afterwards, because I think you'd need his approval for the license change.

This said, I can't tell when I'll take the time to work on that...

**Comment by timakro** [comment]
@lafrech Ah, now I understand you. I guess that could be a problem, I will keep that in mind.

**Pull request merged**